Fix slicer upload in some cases

smallhive opened the pull request:
The case when the reader returns io.EOF with the last data chunk. Slicer was closing the writer without payload writing

carpawell reviewed (commented):
lol. Does it mean that RC10 isn't perfect again in term of `Slicer`? How do you found it?

Commit's body does not end with a period.

smallhive commented:
> lol. Does it mean that RC10 isn't perfect again in term of `Slicer`? How do you found it?
> 
> Commit's body does not end with a period.

Period is fixed. I'm updating our gates. Inside HTTP one I updated all upload places to slicer using. 
One test started to fail: debug, investigation, found, fixed, profit :smiley:

roman-khimov reviewed (changes requested):
It's disappointing to see this given that exactly this issue had been discussed numerous times in slicer-related PRs. Yet at the same all PRs were reviewed successfully before the respective merges.

Let's hope the new test won't let it happen again.

cthulhu-rider commented on object/slicer/slicer_test.go:
```diff
@@ -269,6 +269,23 @@ func testSlicer(t *testing.T, size, sizeLimit uint64) {
+	point   int
+}
+
+func (l *eofOnLastChunkReader) Read(p []byte) (int, error) {
```
lets also return `0, nil` to test all `io.Reader` doc statements? For example 
```go
func (l *eofOnLastChunkReader) Read(p []byte) (int, error) {
  if !l.readEmpty {
    l.readEmpty = true
    return 0, nil
  }
```

smallhive replied:
Extended with new test

cthulhu-rider replied:
does it ever return `0, nil`?

smallhive replied:
No, depending on the condition

cthulhu-rider replied:
lets add then, it's simple (see initial comment) and tests possible `io.Reader` behavior

smallhive replied:
Updated
